grr/server/grr_response_server/flows/general/osquery.py

```diff
@@ -88,5 +166,41 @@ def Process(
     for response in responses:
       self.SendReply(response)
 
+    try:
+      self._FileCollectionFromColumns(responses)
+    except _FileCollectionLimitsExceeded as e:
+      raise flow_base.FlowError(str(e))
+
   def GetProgress(self) -> rdf_osquery.OsqueryProgress:
     return self.state.progress
+
+  def ProcessCollectedFiles(
+      self,
+      responses: flow_responses.Responses[rdf_client_fs.StatEntry],
+  ) -> None:
+    if not responses.success:
+      raise flow_base.FlowError(responses.status)
+
+    for result in responses:
+      self.SendReply(result)
+
+  def GetFilesArchiveMappings(
+      self,
+      flow_results: Iterator[rdf_flow_objects.FlowResult],
+  ) -> Iterator[flow_base.ClientPathArchiveMapping]:
+    for result in flow_results:
+      try:
+        osquery_file = _ExtractFileInfo(result)
+      except _ResultNotRelevantError:
+        continue
+
+      client_path = db.ClientPath.FromPathSpec(
+          self.client_id, osquery_file.stat_entry.pathspec)
+
+      # This approach keeps the client directory structure
+      # TODO(simstoykov): Check whether we should really keep the tree structure
+      # TODO(simstoykov): What checks need to be performed on the full_client_path?
+      full_client_path = osquery_file.stat_entry.pathspec.path
```

Comment:
We should use this to convert the pathspec to a filename:
https://github.com/google/grr/blob/ec1bd364f889e03f26a97f091e7098bc6db8f6f3/grr/server/grr_response_server/rdfvalues/objects.py#L385

PathSpec is a complex data structure, and the function above handles a lot of edge cases correctly.

See, for example (note that it just uses the filename without the path, though):
https://github.com/google/grr/blob/252af5d85f02121a75f8d83ebbee8e3d7e4f5fc9/grr/server/grr_response_server/flows/general/webhistory.py#L428

Reply:
Do you propose we use `ClientPath.FromPathSpec` (as in the `webhistory.py` example) or `PathInfo.FromPathSpec` (the one referenced)? Also what method would give the full path, and what edge cases would that cover? Please note that I use `full_client_path` to construct the `target_path`, not the `client_path`.

Reply:
Good catch - I talk about `PathInfo.FromPathSpec`, but `webhistory.py` has `db.ClientPath.FromPathSpec`. `db.ClientPath.FromPathSpec` is a thin wrapper on top of `PathInfo.FromPathSpec`, so I'd suggest using `PathInfo.FromPathSpec`.

The edge cases it handles are these:
* PathSpec is a recursive data structure. If you just use `pathspec.path`, you ignore the nested component of the pathspec. (see https://github.com/google/grr/blob/ec1bd364f889e03f26a97f091e7098bc6db8f6f3/grr/proto/grr_response_proto/jobs.proto#L612)
* Path elements might have stream names (used in NTFS), see https://github.com/google/grr/blob/ec1bd364f889e03f26a97f091e7098bc6db8f6f3/grr/server/grr_response_server/rdfvalues/objects.py#L394